podpac/core/cache/cache.py

```diff
@@ -0,0 +1,428 @@
+
+class CacheException(Exception):
+    """Summary
+    """
+    pass
+
+
+class CacheCtrl(object):
+
+    def __init__(self, cache_stores=[]):
+        self._cache_stores = cache_stores
+        self._cache_mode = None
+
+    def _determine_mode(self, mode):
+        if mode is None:
+            mode = self._cache_mode
+            if mode is None:
+                mode = 'all'
+        return mode
+
+    def put(self, node, data, key, coordinates=None, mode=None, update=False):
+        '''Cache data for specified node.
+        
+        Parameters
+        ------------
+        node : Node
+            node requesting storage.
+        data : any
+            Data to cache
+        key : str
+            Cached object key, e.g. 'output'.
+        coordinates : Coordinates, optional
+            Coordinates for which cached object should be retrieved, for coordinate-dependent data such as evaluation output
+        mode : str
+            determines what types of the `CacheStore` are affected: 'ram','disk','network','all'. Defaults to `node._cache_mode` or 'all'. Overriden by `self._cache_mode` if `self._cache_mode` is not `None`.
+        update : bool
+            If True existing data in cache will be updated with `data`, If False, error will be thrown if attempting put something into the cache with the same node, key, coordinates of an existing entry.
+        '''
+        mode = self._determine_mode(mode)
+        for c in self._cache_stores:
+            if c.cache_modes_matches(set([mode])):
+                c.put(node=node, data=data, key=key, coordinates=coordinates, update=update)
+        
+
+    def get(self, node, key, coordinates=None, mode=None):
+        '''Get cached data for this node.
+        
+        Parameters
+        ------------
+        node : Node
+            node requesting storage.
+        key : str
+            Cached object key, e.g. 'output'.
+        coordinates : Coordinates, optional
+            Coordinates for which cached object should be retrieved, for coordinate-dependent data such as evaluation output
+        mode : str
+            determines what types of the `CacheStore` are affected: 'ram','disk','network','all'. Defaults to `node._cache_mode` or 'all'. Overriden by `self._cache_mode` if `self._cache_mode` is not `None`.
+            
+        Returns
+        -------
+        data : any
+            The cached data.
+        
+        Raises
+        -------
+        CacheError
+            If the data is not in the cache.
+        '''
+        mode = self._determine_mode(mode)
+        for c in self._cache_stores:
+            if c.cache_modes_matches(set([mode])):
+                if c.has(node=node, key=key, coordinates=coordinates):
+                    return c.get(node=node, key=key, coordinates=coordinates)
+        raise CacheException("Requested data is not in any cache stores.")
+
+    def rem(self, node=None, key=None, coordinates=None, mode=None):
```

Comment:
With the node refactor we went with `del`. Would you mind changing `rem` to `del` so everything matches?

Reply:
I get an error when trying to define a del method for a class in ipython. I think because of the built in python del method:

```
n [1]: class C:
   ...:     def del():
  File "<ipython-input-1-74da07aeaab4>", line 2
    def del():
          ^
SyntaxError: invalid syntax
```

Reply:
Ok, we'll have to fix that in the Node then.